### DISCUSSION TITLE
Managed State

### DISCUSSION THREAD

**Discussion by genericblueprint** [tauri-apps/tauri#4202] [General]
Is it possible to access and mutate the managed state from a Rust function that isn't a `#[tauri::command]`?

**Comment by FabianLars** [comment]
Yes but you need access to a struct that implements the Manager trait, so App, AppHandle or Window. Then you can use [state()](https://docs.rs/tauri/1.0.0-rc.0/tauri/trait.Manager.html#method.state) to get managed state
```rs
let my_state: tauri::State<MyState> = app.state();
// or
let my_state = app.state::<MyState>();
```

**Comment by genericblueprint** [comment]
How would I access something like `app` in a separate module? Is there a globally accessible instance of `app`? Perhaps I am overthinking this.

**Reply by FabianLars** [reply]
> Perhaps I am overthinking this.

Probably :P Many people do.

> Is there a globally accessible instance of app?

Nope, you somehow need to move it from the setup hook or tauri commands. As a last resort you can create a global-like AppHandle with the once_cell crate.

**Reply by genericblueprint** [reply]
Ok, using once_cell or lazy_static were where I was heading naturally. I was really trying to avoid any "singleton" type for obvious reasons, but for what I'm trying to accomplish it may be appropriate. I'm sure programmers everywhere will yell at me.

Thank you for your help!
